Question: Dynamic registry and loading from proto sources

First of all I do apologise if I'm missing something very obvious or if I am misunderstanding or abusing things in this wonderful package.

I'm trying to setup a dynamic registry with a fetcher that is able to load protos from their source files without much success.

I've managed to get the dynamic registry to work when the fetcher manually constructs and returns a `*ptype.Type` but I'm struggling to figure out how to convert the result from the `protoparse.Parser` into a `*ptype.Type`. It seems to be dying trying to get the options, but I'm probably doing something very wrong here.

ps. Just a weird side-note, I cannot for the life of me debug this test using delve, it just dies with `Failed to produce stack trace!` -- will need to dig into this at some point :D

Thank you very much in advance for your help and also for this amazing package.
George

---

### Testcase

```golang
package dynamic

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"testing"

	"github.com/davecgh/go-spew/spew"
	"github.com/golang/protobuf/proto"
	"github.com/jhump/protoreflect/desc/protoparse"
	"github.com/jhump/protoreflect/dynamic"
	"github.com/jhump/protoreflect/dynamic/msgregistry"
	"github.com/stretchr/testify/assert"
)

func TestSomethingWeird2(t *testing.T) {
	protoSource := `
		syntax = "proto2";
		package foo;
		message Bar {
			optional string abc = 1;
			optional int32 def = 2;
		}`

	registry := (&msgregistry.MessageRegistry{}).
		WithMessageFactory(dynamic.NewMessageFactoryWithDefaults())

	reader := func(filename string) (io.ReadCloser, error) {
		if filename != "https://test/foo.Bar" {
			return nil, errors.New("not found")
		}
		return ioutil.NopCloser(bytes.NewReader([]byte(protoSource))), nil
	}

	parser := &protoparse.Parser{
		Accessor: reader,
	}

	fetcher := func(url string, enum bool) (proto.Message, error) {
		fmt.Println(url)
		if url != "https://test/foo.Bar" {
			return nil, errors.New("not found")
		}

		fds, err := parser.ParseFiles(url)
		if err != nil {
			return nil, err
		}

		if len(fds) == 0 {
			return nil, errors.New("missing descs")
		}

		fooMessage := fds[0].FindMessage("foo.Bar")
		spew.Dump(fooMessage)

		fooPType := registry.MessageAsPType(fooMessage)
		spew.Dump(fooPType)

		return fooPType, nil
	}

	registry = registry.WithFetcher(fetcher)

	md, err := registry.FindMessageTypeByUrl("test/foo.Bar")
	assert.NoError(t, err)
	assert.NotNil(t, md)

	if md == nil {
		t.FailNow()
	}

	dm := dynamic.NewMessageFactoryWithDefaults().NewMessage(md).(*dynamic.Message)
	dm.SetFieldByNumber(1, "fubar")
	dm.SetFieldByNumber(2, int32(123))
	dm.SetFieldByNumber(1, "snafu")
	dm.SetFieldByNumber(2, int32(456))

	js, err := dm.MarshalJSON()
	assert.NoError(t, err)
	fmt.Println(string(js))
}
```

### Result

```
=== RUN   TestSomethingWeird
(*desc.MessageDescriptor)(0xc0001ebb20)(name:"Bar" field:<name:"abc" number:1 label:LABEL_OPTIONAL type:TYPE_STRING json_name:"abc" > field:<name:"def" number:2 label:LABEL_OPTIONAL type:TYPE_INT32 json_name:"def" > )
--- FAIL: TestSomethingWeird2 (0.00s)
panic: reflect: call of reflect.Value.Type on zero Value [recovered]
	panic: reflect: call of reflect.Value.Type on zero Value

goroutine 19 [running]:
testing.tRunner.func1(0xc00014cb00)
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:792 +0x387
panic(0x1436720, 0xc0002184c0)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
reflect.Value.Type(0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.11/libexec/src/reflect/value.go:1713 +0x16c
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).options(0xc000149300, 0x1533bc0, 0x0, 0x10, 0x10, 0x149bea0)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:596 +0x10c
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).fieldAsPType(0xc000149300, 0xc00020da40, 0x2)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:456 +0x69
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).MessageAsPType(0xc000149300, 0xc0001ebb20, 0x1)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:438 +0xcb
command-line-arguments.TestSomethingWeird2.func2(0xc0001e91c0, 0x14, 0x0, 0xc0001cef60, 0x30, 0x30, 0x1496840)
	/Users/geoah/src/nope.io/go/playground/dynamic/dynamic3_test.go:62 +0x33c
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*resolutionContext).addType(0xc000153dc0, 0xc0001e91c0, 0x14, 0xc0000bf9e0, 0x187b700, 0x0, 0x0)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/ptype_resolver.go:445 +0xbb
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*typeResolver).resolveUrlToMessageDescriptor(0xc000149308, 0xc0001e91c0, 0x14, 0x0, 0x0, 0x0)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/ptype_resolver.go:282 +0x278
nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry.(*MessageRegistry).FindMessageTypeByUrl(0xc000149300, 0x14b7940, 0xc, 0x105e370, 0xc00004c720, 0xe4b2a72b)
	/Users/geoah/src/nope.io/go/playground/dynamic/vendor/github.com/jhump/protoreflect/dynamic/msgregistry/message_registry.go:165 +0x91
command-line-arguments.TestSomethingWeird2(0xc00014cb00)
	/Users/geoah/src/nope.io/go/playground/dynamic/dynamic3_test.go:70 +0x21d
testing.tRunner(0xc00014cb00, 0x14cebd0)
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:878 +0x353
FAIL	command-line-arguments	0.018s
```

--- a/dynamic/msgregistry/message_registry.go
+++ b/dynamic/msgregistry/message_registry.go
@@ -590,10 +590,16 @@ func (r *MessageRegistry) methodAsApi(md *desc.MethodDescriptor) *api.Method {
 func (r *MessageRegistry) options(options proto.Message) []*ptype.Option {
 	rv := reflect.ValueOf(options)
 	if rv.Kind() == reflect.Ptr {
+		if rv.IsNil() {
+			return nil
+		}
 		rv = rv.Elem()
 	}
 	var opts []*ptype.Option
 	for _, p := range proto.GetProperties(rv.Type()).Prop {
+		if p.Tag == 0 {
+			continue
+		}
 		o := r.option(p.OrigName, rv.FieldByName(p.Name))
 		if o != nil {
 			opts = append(opts, o...)
@@ -619,12 +625,21 @@ func (r *MessageRegistry) option(name string, value reflect.Value) []*ptype.Opti
 	if value.Kind() == reflect.Slice && value.Type() != typeOfBytes {
 		// repeated field
 		ret := make([]*ptype.Option, value.Len())
+		j := 0
 		for i := 0; i < value.Len(); i++ {
-			ret[i] = r.singleOption(name, value.Index(i))
+			opt := r.singleOption(name, value.Index(i))
+			if opt != nil {
+				ret[j] = opt
+				j++
+			}
 		}
-		return ret
+		return ret[:j]
 	} else {
-		return []*ptype.Option{r.singleOption(name, value)}
+		opt := r.singleOption(name, value)
+		if opt != nil {
+			return []*ptype.Option{opt}
+		}
+		return nil
 	}
 }
 
@@ -647,12 +662,21 @@ func wrap(v reflect.Value) proto.Message {
 	if pm, ok := v.Interface().(proto.Message); ok {
 		return pm
 	}
+	if !v.IsValid() {
+		return nil
+	}
+	if v.Kind() == reflect.Ptr {
+		if v.IsNil() {
+			return nil
+		}
+		v = v.Elem()
+	}
 	switch v.Kind() {
 	case reflect.Bool:
 		return &wrappers.BoolValue{Value: v.Bool()}
 	case reflect.Slice:
 		if v.Type() != typeOfBytes {
-			return nil
+			panic(fmt.Sprintf("cannot convert/wrap %T as proto", v.Type()))
 		}
 		return &wrappers.BytesValue{Value: v.Bytes()}
 	case reflect.String:
@@ -670,7 +694,7 @@ func wrap(v reflect.Value) proto.Message {
 	case reflect.Uint64:
 		return &wrappers.UInt64Value{Value: v.Uint()}
 	default:
-		return nil
+		panic(fmt.Sprintf("cannot convert/wrap %T as proto", v.Type()))
 	}
 }
 

--- a/dynamic/msgregistry/message_registry_test.go
+++ b/dynamic/msgregistry/message_registry_test.go
@@ -2,6 +2,9 @@ package msgregistry
 
 import (
 	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
 	"reflect"
 	"strings"
 	"testing"
@@ -17,6 +20,7 @@ import (
 	"google.golang.org/genproto/protobuf/source_context"
 
 	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/desc/protoparse"
 	"github.com/jhump/protoreflect/dynamic"
 	"github.com/jhump/protoreflect/internal/testprotos"
 	"github.com/jhump/protoreflect/internal/testutil"
@@ -704,7 +708,87 @@ func TestMessageRegistry_MarshalAndUnmarshalAny(t *testing.T) {
 }
 
 func TestMessageRegistry_MessageDescriptorToPType(t *testing.T) {
-	// TODO
+	protoSource := `
+		syntax = "proto2";
+		package foo;
+		message Bar {
+			optional string abc = 1 [deprecated = true];
+			repeated int32 def = 2 [packed = true];
+			optional string ghi = 3 [default = "foobar"];
+			oneof oo {
+				uint64 nid = 4;
+				string sid = 5;
+			}
+		}`
+	p := protoparse.Parser{
+		Accessor: func(filename string) (io.ReadCloser, error) {
+			if filename == "test.proto" {
+				return ioutil.NopCloser(strings.NewReader(protoSource)), nil
+			}
+			return nil, os.ErrNotExist
+		},
+	}
+	fds, err := p.ParseFiles("test.proto")
+	testutil.Ok(t, err)
+	fd := fds[0]
+
+	msg := NewMessageRegistryWithDefaults().MessageAsPType(fd.GetMessageTypes()[0])
+
+	// quick check of the resulting message's properties
+	testutil.Eq(t, "foo.Bar", msg.Name)
+	testutil.Eq(t, []string{"oo"}, msg.Oneofs)
+	testutil.Eq(t, ptype.Syntax_SYNTAX_PROTO2, msg.Syntax)
+	testutil.Eq(t, "test.proto", msg.SourceContext.GetFileName())
+	testutil.Eq(t, 0, len(msg.Options))
+
+	testutil.Eq(t, "abc", msg.Fields[0].Name)
+	testutil.Eq(t, ptype.Field_CARDINALITY_OPTIONAL, msg.Fields[0].Cardinality)
+	testutil.Eq(t, ptype.Field_TYPE_STRING, msg.Fields[0].Kind)
+	testutil.Eq(t, "", msg.Fields[0].DefaultValue)
+	testutil.Eq(t, int32(1), msg.Fields[0].Number)
+	testutil.Eq(t, int32(0), msg.Fields[0].OneofIndex)
+	testutil.Eq(t, 1, len(msg.Fields[0].Options))
+	testutil.Eq(t, "deprecated", msg.Fields[0].Options[0].Name)
+	// make sure the value is a wrapped bool
+	var v ptypes.DynamicAny
+	err = ptypes.UnmarshalAny(msg.Fields[0].Options[0].Value, &v)
+	testutil.Ok(t, err)
+	testutil.Ceq(t, &wrappers.BoolValue{Value: true}, v.Message, func(a, b interface{}) bool {
+		return proto.Equal(a.(proto.Message), b.(proto.Message))
+	})
+
+	testutil.Eq(t, "def", msg.Fields[1].Name)
+	testutil.Eq(t, ptype.Field_CARDINALITY_REPEATED, msg.Fields[1].Cardinality)
+	testutil.Eq(t, ptype.Field_TYPE_INT32, msg.Fields[1].Kind)
+	testutil.Eq(t, "", msg.Fields[1].DefaultValue)
+	testutil.Eq(t, int32(2), msg.Fields[1].Number)
+	testutil.Eq(t, int32(0), msg.Fields[1].OneofIndex)
+	testutil.Eq(t, true, msg.Fields[1].Packed)
+	testutil.Eq(t, 0, len(msg.Fields[1].Options))
+
+	testutil.Eq(t, "ghi", msg.Fields[2].Name)
+	testutil.Eq(t, ptype.Field_CARDINALITY_OPTIONAL, msg.Fields[2].Cardinality)
+	testutil.Eq(t, ptype.Field_TYPE_STRING, msg.Fields[2].Kind)
+	testutil.Eq(t, "foobar", msg.Fields[2].DefaultValue)
+	testutil.Eq(t, int32(3), msg.Fields[2].Number)
+	testutil.Eq(t, int32(0), msg.Fields[2].OneofIndex)
+	testutil.Eq(t, 0, len(msg.Fields[2].Options))
+
+	testutil.Eq(t, "nid", msg.Fields[3].Name)
+	testutil.Eq(t, ptype.Field_CARDINALITY_OPTIONAL, msg.Fields[3].Cardinality)
+	testutil.Eq(t, ptype.Field_TYPE_UINT64, msg.Fields[3].Kind)
+	testutil.Eq(t, "", msg.Fields[3].DefaultValue)
+	testutil.Eq(t, int32(4), msg.Fields[3].Number)
+	testutil.Eq(t, int32(1), msg.Fields[3].OneofIndex)
+	testutil.Eq(t, 0, len(msg.Fields[3].Options))
+
+	testutil.Eq(t, "sid", msg.Fields[4].Name)
+	testutil.Eq(t, ptype.Field_CARDINALITY_OPTIONAL, msg.Fields[4].Cardinality)
+	testutil.Eq(t, ptype.Field_TYPE_STRING, msg.Fields[4].Kind)
+	testutil.Eq(t, "", msg.Fields[4].DefaultValue)
+	testutil.Eq(t, int32(5), msg.Fields[4].Number)
+	testutil.Eq(t, int32(1), msg.Fields[4].OneofIndex)
+	testutil.Eq(t, 0, len(msg.Fields[4].Options))
 }
 
 func TestMessageRegistry_EnumDescriptorToPType(t *testing.T) {

